src/sdf/include/sdf.h

```diff
@@ -0,0 +1,474 @@
+      glm::ivec4 neighborIndex = gridIndex + Neighbors(i);
+      if (neighborIndex.w == 2) {
+        neighborIndex += 1;
+        neighborIndex.w = 0;
+      }
+      const glm::vec3 iPos = Position(neighborIndex);
+      const float val = BoundedSDF(neighborIndex);
+      if (val * gridVert.distance > 0 || (val == 0 && gridVert.distance == 0))
+        continue;
+      keep = true;
+
+      // Record the nearest intersection of all 14 edges, only if it is close
+      // enough to allow this gridVert to safely move to it without inverting
+      // any tetrahedra.
+      const glm::vec3 interp = (val * position - gridVert.distance * iPos) /
+                               (val - gridVert.distance);
+      const glm::vec3 delta = interp - position;
+      const float dist2 = glm::dot(delta, delta);
+      if (dist2 < minDist2) {
+        // gridVert.edgeIndex = i;
+        minDist2 = dist2;
+      }
+
+      // These seven edges are uniquely owned by this gridVert; any of them
+      // which intersect the surface create a vert.
+      if (i < 7) {
+        const int idx = AtomicAdd(*vertIndex, 1);
+        vertPos[idx] = interp;
+        gridVert.edgeVerts[i] = idx;
+      }
+    }
+
+    if (keep && gridVerts.Insert(gridVert)) printf("out of space!\n");
+  }
+};
+
+struct BuildTris {
+  glm::ivec3* triVerts;
+  int* triIndex;
+  const HashTableD gridVerts;
+
+  __host__ __device__ void CreateTri(const glm::ivec3& tri,
+                                     const int edges[6]) {
+    if (tri[0] < 0) return;
+    int idx = AtomicAdd(*triIndex, 1);
+    triVerts[idx] = {edges[tri[0]], edges[tri[1]], edges[tri[2]]};
+  }
+
+  __host__ __device__ void CreateTris(const glm::ivec4& tet,
+                                      const int edges[6]) {
+    const int i = (tet[0] > 0 ? 1 : 0) + (tet[1] > 0 ? 2 : 0) +
+                  (tet[2] > 0 ? 4 : 0) + (tet[3] > 0 ? 8 : 0);
+    CreateTri(TetTri0(i), edges);
+    CreateTri(TetTri1(i), edges);
+  }
+
+  __host__ __device__ void operator()(int idx) {
+    const GridVert& base = gridVerts.At(idx);
+    if (base.key == kOpen) return;
+
+    const glm::ivec4 baseIndex = DecodeMorton(base.key);
+
+    glm::ivec4 leadIndex = baseIndex;
+    if (leadIndex.w == 0)
+      leadIndex.w = 1;
+    else {
+      leadIndex += 1;
+      leadIndex.w = 0;
+    }
+
+    const GridVert& leadVert = gridVerts[MortonCode(leadIndex)];
+    if (leadVert.key == kOpen) return;
+
+    // This GridVert is in charge of the 6 tetrahedra surrounding its edge in
+    // the (1,1,1) direction, attached to leadVert.
+    glm::ivec4 tet(leadVert.Inside(), base.Inside(), -2, -2);
+    glm::ivec4 thisIndex = baseIndex;
+    thisIndex.x += 1;
+
+    GridVert thisVert = gridVerts[MortonCode(thisIndex)];
+    bool skipTet = thisVert.key == kOpen;
+
+    tet[2] = thisVert.Inside();
+    int edges[6] = {base.edgeVerts[0], -1, -1, -1, -1, -1};
+    for (const int i : {0, 1, 2}) {
+      edges[1] = base.edgeVerts[i + 1];
```

Comment:
This part is relatively complicated, maybe we can reorder some statements and avoid reusing the array to make it more readable?

```c++
for (const int i : {0, 1, 2}) {
  thisIndex = leadIndex;
  thisIndex[Prev3(i)] -= 1;
  GridVert nextVert = gridVerts[MortonCode(thisIndex)];
  tet[3] = nextVert.Inside();

  int edges1[6] = {
    base.edgeVerts[0],
    base.edgeVerts[i+1],
    nextVert.edgeVerts[Next3(i) + 4],
    nextVert.edgeVerts[Prev3(i) + 1],
    thisVert.edgeVerts[i + 4],
    base.edgeVerts[Prev3(i) + 4]
  };
  thisVert = nextVert;
  if (!skipTet && thisVert.key != kOpen) CreateTris(tet, edges1);
  skipTet = thisVert.key == kOpen;
  tet[2] = tet[3];
  int edges2[6] = ...
}
```

Reply:
(I personally think that it is easy to get lost reading many array element operations, especially with some order dependencies with other variables)

Reply:
Okay, this has been cleaned up a lot; thanks for the suggestion, I ended up using that and a few other things too.